Manually specify scryptenc parameters

gperciva opened the pull request.

gperciva commented:
This allows us to specify logN, r, p on the command line, as per #200.

I plan to introduce this as 2 or 3 separate PRs; the first of which is #250.

gperciva commented:
Ready for review.

Big question mark about bounds on `--logN`, `-r`, `-p`.  I think it's reasonable to have some (in addition the hard limits of the scrypt format), but I'm wildly guessing at what would be appropriate.

cperciva commented on main.c:
```diff
@@ -62,6 +62,21 @@ usage(void)
 	exit(1);
 }
 
+#define GETOPT_PARSENUM_WITHIN(var, min, max)				\
```
We normally write these as
```
#define foo(bar) do { \
        blah blah blah; \
} while (0)
```

cperciva replied:
That is to say, with the `do {` on the opening line and with a level less indentation than you have here.

cperciva commented on main.c:
```diff
@@ -160,6 +175,9 @@ main(int argc, char *argv[])
 			force_resources = 1;
 			break;
+		GETOPT_OPTARG("--logN"):
+			GETOPT_PARSENUM_WITHIN(&params.logN, 10, 32);
```
I'd be inclined to make this range 10--40, just for future-proofing.

We're applying the "make sure we have a reasonable amount of memory/time" tests (assuming no `-f`) right?  So there should already be a seatbelt to protect against crazy parameters.

gperciva replied:
No, there's no "reasonable" checks.  `scryptenc_setup:`

```
        /* Determine parameters. */
        if (P->logN != 0) {
                /* Check logN, r, p, without any resource limits. */
                if ((rc = checkparams(0, 0, 0, P->logN, P->r, P->p,
                    verbose, 1)) != 0) {
```

The idea is that if somebody explicitly specifies N, r, p, then we assume that they know what they're doing.

(this change was in a previous PR, in case you're wondering why you didn't see that code in this PR)

gperciva replied:
* the "without any resource limits" refers to the final `1` in `checkparams()`, which enables `force`.

gperciva commented:
Rebased, ready for review.

cperciva commented on main.c:
```diff
@@ -54,14 +54,26 @@ usage(void)
+#define GETOPT_PARSENUM_WITHIN_UNSIGNED(var, min, max) do {		\
+	if (PARSENUM((var), optarg, (min), (max))) {			\
+		if (errno == ERANGE) {					\
+			warn0("%s must be between %zu and %zu"		\
```
You're printing these as `size_t` variables, but you're not casting them to that type.

gperciva replied:
Argh.

I'm tempted to do a whole parsenum-style "print as double, intmax_t, or uintmax_t" macro, but I can't find enough instances of "parse a number within this range" in our code to make it IMO worth it.

Since unattached numbers are intepreted as `int`, I renamed this to `GETOPT_PARSENUM_WITHIN_INT`.

cperciva replied:
Well, I guess you could do that.  Although I would still prefer to see explicit casts inside the printf.

But I was more thinking of casting to `uintmax_t` and using `%ju`...

gperciva replied:
That was my first thought, but then I figured that you'd say it was over-engineered.

... ok, updated.

cperciva commented:
In general, macros *should* be over-engineered, since they're inherently somewhat opaque; it's easy for someone using one in the future to not be aware of their limitations.

LGTM, please rebase.

gperciva commented:
Rebased.

I also fixed a few other things:
- renamed the test to `09-explicit-params.sh` (it previously had `manual` instead of `explicit`)
- fixed the alphabetical order of `[-p value]` in the man page.
- tweaked the man-page option text to say:

  > If such explicit parameters are given, the resource limits set by -M, -m, and -t are not enforced.

  (Previously it just said that -M, -m, and -t were ignored)

cperciva commented:
What happens if someone specifies `--logN 15` but does not specify values for *r* and *p*?

gperciva commented:
Assertion error, as desired?

... oh right, we should have a more polite message in `main.c`.

gperciva commented:
Added the checks for attempting to partially set explicit params.

At first, I combined it all into two lines (like the assertions), but I found that unpleasantly complicated.  More importantly, I couldn't find a good wording for the error message that would make sense to people who didn't know what was happening without it exceeding 80 characters.

So instead, I split into 3 checks.  That meant the error message could mirror the man-page sentences, and vastly simplified the `if` statements.

cperciva commented:
LGTM.

gperciva commented:
Rebased, ready for merge.
